File: rust/main/features/monitor/tests/monitor_int_test.rs
//! Integration tests for disk-cleaner-monitor crate.
//! Tests history persistence and process scanning.

use std::fs;

use disk_cleaner_monitor::{DefaultHistoryStore, HistoryEntry, HistoryStore};

// ─── History Persistence ─────────────────────────────────────────────────────

/// @covers: DefaultHistoryStore::record, DefaultHistoryStore::load
#[test]
fn test_history_record_and_load() {
    let tmp = tempfile::tempdir().unwrap();
    let store = DefaultHistoryStore::new(tmp.path());

    store.record(HistoryEntry {
        timestamp: "2026-03-24T10:00:00+00:00".to_string(),
        command: "clean".to_string(),
        profiles: "rust".to_string(),
        projects: 5,
        size_bytes: 1048576,
        size_formatted: "1.00 MiB".to_string(),
        path: "/projects".to_string(),
    });

    let entries = store.load();
    assert_eq!(entries.len(), 1);
    assert_eq!(entries[0].command, "clean");
    assert_eq!(entries[0].projects, 5);
    assert_eq!(entries[0].size_bytes, 1048576);
}

/// @covers: DefaultHistoryStore::record (multiple entries)
#[test]
fn test_history_appends_entries() {
    let tmp = tempfile::tempdir().unwrap();
    let store = DefaultHistoryStore::new(tmp.path());

    for i in 0..5 {
        store.record(HistoryEntry {
            timestamp: String::new(),
            command: "clean".to_string(),
            profiles: "rust".to_string(),
            projects: i,
            size_bytes: (i as i64) * 1024,
            size_formatted: format!("{} B", i * 1024),
            path: "/projects".to_string(),
        });
    }

    let entries = store.load();
    assert_eq!(entries.len(), 5);
    assert_eq!(entries[4].projects, 4);
}

/// @covers: DefaultHistoryStore::record (max 100 entries)
#[test]
fn test_history_keeps_max_100_entries() {
    let tmp = tempfile::tempdir().unwrap();
    let store = DefaultHistoryStore::new(tmp.path());

    for i in 0..110 {
        store.record(HistoryEntry {
            timestamp: String::new(),
            command: "clean".to_string(),
            profiles: "rust".to_string(),
            projects: i,
            size_bytes: 0,
            size_formatted: "0 B".to_string(),
            path: "/projects".to_string(),
        });
    }

    let entries = store.load();
    assert_eq!(entries.len(), 100);
    // First entry should be #10 (oldest 10 dropped)
    assert_eq!(entries[0].projects, 10);
}

/// @covers: DefaultHistoryStore::load (empty file)
#[test]
fn test_history_load_empty_returns_empty() {
    let tmp = tempfile::tempdir().unwrap();
    let store = DefaultHistoryStore::new(tmp.path());

    let entries = store.load();
    assert!(entries.is_empty());
}

/// @covers: DefaultHistoryStore::load (corrupt file)
#[test]
fn test_history_load_corrupt_file_returns_empty() {
    let tmp = tempfile::tempdir().unwrap();
    fs::write(tmp.path().join("history.json"), "not valid json{{{").unwrap();

    let store = DefaultHistoryStore::new(tmp.path());
    let entries = store.load();
    assert!(entries.is_empty());
}

/// @covers: DefaultHistoryStore::record (auto-timestamps)
#[test]
fn test_history_record_auto_timestamps_empty_timestamp() {
    let tmp = tempfile::tempdir().unwrap();
    let store = DefaultHistoryStore::new(tmp.path());

    store.record(HistoryEntry {
        timestamp: String::new(), // should be auto-filled
        command: "analyze".to_string(),
        profiles: "node".to_string(),
        projects: 3,
        size_bytes: 2048,
        size_formatted: "2.00 KiB".to_string(),
        path: "/data".to_string(),
    });

    let entries = store.load();
    assert_eq!(entries.len(), 1);
    assert!(!entries[0].timestamp.is_empty(), "Timestamp should be auto-filled");
}

// ─── Monitor Run (smoke test) ───────────────────────────────────────────────

/// @covers: monitor::run (does not panic)
#[test]
fn test_monitor_run_history_only_does_not_panic() {
    let tmp = tempfile::tempdir().unwrap();
    let scanner = disk_cleaner_config::create_scanner(
        tmp.path().to_path_buf(),
        vec![],
        vec![],
        false,
        false,
    );

    // Should not panic with empty history
    disk_cleaner_monitor::run(&scanner, tmp.path(), true);
}

/// @covers: monitor::run (full mode)
#[test]
fn test_monitor_run_full_does_not_panic() {
    let tmp = tempfile::tempdir().unwrap();
    let scanner = disk_cleaner_config::create_scanner(
        tmp.path().to_path_buf(),
        vec![],
        vec![],
        false,
        false,
    );

    // Full mode: process scan + history
    disk_cleaner_monitor::run(&scanner, tmp.path(), false);
}

File: java/tests/monitor/MonitorIntTest.java
package tests.monitor;

import main.features.config.saf.ConfigFactory;
import main.features.monitor.saf.MonitorFactory;
import main.features.monitor.api.HistoryEntry;
import main.features.monitor.core.DefaultHistoryStore;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

/** Integration tests for monitor feature. */
public class MonitorIntTest {

    static int passed = 0;
    static int failed = 0;

    public static void main(String[] args) {
        System.out.println("=== Monitor Integration Tests ===");
        System.out.println();

        testHistoryRecordAndLoad();
        testHistoryAppendsEntries();
        testHistoryKeepsMax100();
        testHistoryLoadEmptyReturnsEmpty();
        testHistoryLoadCorruptReturnsEmpty();
        testHistoryAutoTimestamps();
        testMonitorRunHistoryOnlyDoesNotPanic();
        testMonitorRunFullDoesNotPanic();

        System.out.println();
        System.out.println("Results: " + passed + " passed, " + failed + " failed");
        if (failed > 0) System.exit(1);
    }

    static String createTempDir() { var d = System.getProperty("java.io.tmpdir") + File.separator + "dc-mon-" + System.nanoTime(); new File(d).mkdirs(); return d; }
    static void deleteRecursive(File f) { if (f.isDirectory()) { var ch = f.listFiles(); if (ch != null) for (var c : ch) deleteRecursive(c); } f.delete(); }
    static void check(String name, boolean cond) { if (cond) { passed++; System.out.println("  PASS  " + name); } else { failed++; System.out.println("  FAIL  " + name); } }

    static void testHistoryRecordAndLoad() {
        var dir = createTempDir();
        try {
            var store = new DefaultHistoryStore(dir);
            var entry = new HistoryEntry();
            entry.timestamp = "2026-03-24T10:00:00+00:00";
            entry.command = "clean";
            entry.profiles = "rust";
            entry.projects = 5;
            entry.sizeBytes = 1048576;
            entry.sizeFormatted = "1.00 MiB";
            entry.path = "/projects";
            store.record(entry);
            var entries = store.load();
            check("test_history_record_and_load",
                    entries.length == 1 && entries[0].command.equals("clean") && entries[0].projects == 5);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testHistoryAppendsEntries() {
        var dir = createTempDir();
        try {
            var store = new DefaultHistoryStore(dir);
            for (int i = 0; i < 5; i++) {
                var e = new HistoryEntry();
                e.timestamp = ""; e.command = "clean"; e.profiles = "rust";
                e.projects = i; e.sizeBytes = i * 1024; e.sizeFormatted = i + " B"; e.path = "/p";
                store.record(e);
            }
            var entries = store.load();
            check("test_history_appends_entries", entries.length == 5 && entries[4].projects == 4);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testHistoryKeepsMax100() {
        var dir = createTempDir();
        try {
            var store = new DefaultHistoryStore(dir);
            for (int i = 0; i < 110; i++) {
                var e = new HistoryEntry();
                e.timestamp = ""; e.command = "clean"; e.profiles = "rust";
                e.projects = i; e.sizeBytes = 0; e.sizeFormatted = "0 B"; e.path = "/p";
                store.record(e);
            }
            var entries = store.load();
            check("test_history_keeps_max_100", entries.length == 100 && entries[0].projects == 10);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testHistoryLoadEmptyReturnsEmpty() {
        var dir = createTempDir();
        try {
            var store = new DefaultHistoryStore(dir);
            check("test_history_load_empty_returns_empty", store.load().length == 0);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testHistoryLoadCorruptReturnsEmpty() {
        var dir = createTempDir();
        try {
            var w = new FileWriter(dir + File.separator + "history.json");
            w.write("not valid json{{{");
            w.close();
            var store = new DefaultHistoryStore(dir);
            check("test_history_load_corrupt_returns_empty", store.load().length == 0);
        } catch (IOException e) { check("test_history_load_corrupt_returns_empty", false); }
        finally { deleteRecursive(new File(dir)); }
    }

    static void testHistoryAutoTimestamps() {
        var dir = createTempDir();
        try {
            var store = new DefaultHistoryStore(dir);
            var e = new HistoryEntry();
            e.timestamp = ""; e.command = "analyze"; e.profiles = "node";
            e.projects = 3; e.sizeBytes = 2048; e.sizeFormatted = "2.00 KiB"; e.path = "/data";
            store.record(e);
            var entries = store.load();
            check("test_history_auto_timestamps",
                    entries.length == 1 && !entries[0].timestamp.isEmpty());
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testMonitorRunHistoryOnlyDoesNotPanic() {
        var dir = createTempDir();
        try {
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[0], false, false);
            MonitorFactory.run(scanner, dir, true);
            check("test_monitor_run_history_only_does_not_panic", true);
        } finally { deleteRecursive(new File(dir)); }
    }

    static void testMonitorRunFullDoesNotPanic() {
        var dir = createTempDir();
        try {
            var scanner = ConfigFactory.createScanner(dir, new String[0], new String[0], false, false);
            MonitorFactory.run(scanner, dir, false);
            check("test_monitor_run_full_does_not_panic", true);
        } finally { deleteRecursive(new File(dir)); }
    }
}
